File: desktop/launcher/AppPaths.cs
namespace LogExplorer.Launcher;

/// <summary>
/// Legacy Remediation Slice 9 §J: every mutable, per-user path the
/// launcher touches lives under <c>%LOCALAPPDATA%\LogExplorer</c> -
/// never inside the installed Program Files tree (which an installed
/// application cannot reliably write to, and never should). Currently
/// this is the backend's own log file and the backend's classification
/// rules configuration (<see cref="DataDirectory"/>). The backend still
/// has no local database (CLAUDE.md §8 - "no application database" is a
/// whole-project invariant, not just the server-side scope): the rules
/// are a small JSON configuration file, never log data.
/// </summary>
internal static class AppPaths
{
    public static string LocalAppDataRoot { get; } =
        Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "LogExplorer");

    public static string LogsDirectory { get; } = Path.Combine(LocalAppDataRoot, "logs");

    /// <summary>
    /// Passed to the backend as <c>LOGEXPLORER_DATA_DIR</c>; the backend
    /// writes <c>classification-rules.json</c> (plus its <c>.bak</c>) here
    /// on the first rule save and creates the directory itself. Deliberately
    /// outside the per-user install directory
    /// (<c>%LOCALAPPDATA%\Programs\Log Explorer</c>, replaced on upgrade and
    /// removed on uninstall), so rules survive upgrades and reinstalls.
    /// </summary>
    public static string DataDirectory { get; } = Path.Combine(LocalAppDataRoot, "data");

    public static string BackendLogFile { get; } = Path.Combine(LogsDirectory, "backend.log");

    public static void EnsureDirectoriesExist()
    {
        Directory.CreateDirectory(LogsDirectory);
    }
}

File: desktop/launcher-macos/src/main/java/com/logexplorer/desktoplauncher/AppPaths.java
package com.logexplorer.desktoplauncher;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

/**
 * REL-1 (mirrors desktop/launcher/AppPaths.cs): every mutable, per-user
 * path this launcher touches lives under
 * {@code ~/Library/Application Support/LogExplorer} - never inside the
 * installed .app bundle (which a packaged application should not write
 * to). Currently this is the backend's own log file, the backend's
 * classification rules configuration ({@link #DATA_DIRECTORY}), the
 * launcher's own single-instance lock file, and a small file recording the
 * port the running instance is bound to (so a second launch can open the
 * browser to the correct address instead of guessing). Still no local
 * database: the rules are a small JSON configuration file, never log data.
 */
final class AppPaths {

    static final Path LOCAL_APP_DATA_ROOT =
        Path.of(System.getProperty("user.home"), "Library", "Application Support", "LogExplorer");

    static final Path LOGS_DIRECTORY = LOCAL_APP_DATA_ROOT.resolve("logs");

    /**
     * Passed to the backend as {@code LOGEXPLORER_DATA_DIR}; the backend
     * writes {@code classification-rules.json} (plus its {@code .bak}) here
     * on the first rule save and creates the directory itself. Outside the
     * .app bundle, so replacing or deleting the bundle (upgrade/uninstall)
     * keeps the user's rules.
     */
    static final Path DATA_DIRECTORY = LOCAL_APP_DATA_ROOT.resolve("data");

    static final Path BACKEND_LOG_FILE = LOGS_DIRECTORY.resolve("backend.log");

    static final Path INSTANCE_LOCK_FILE = LOCAL_APP_DATA_ROOT.resolve("instance.lock");

    static final Path RUNNING_PORT_FILE = LOCAL_APP_DATA_ROOT.resolve("running-port.txt");

    private AppPaths() {
    }

    static void ensureDirectoriesExist() throws IOException {
        Files.createDirectories(LOGS_DIRECTORY);
    }
}
